File: ios/MeshChatSdk.swift
import Foundation
import CoreBluetooth
import UIKit
import UserNotifications

/// Сообщение mesh-чата, отдаваемое наверх (в React Native).
/// Адресация по строковому peerID — как в Android-SDK.
public struct MeshChatMessage {
    public let id: String
    public let senderPeerID: String?
    public let sender: String
    public let content: String
    public let timestamp: TimeInterval   // мс с эпохи
    public let isPrivate: Bool
    public let isRelay: Bool
}

public protocol MeshChatListener: AnyObject {
    func onMessageReceived(_ message: MeshChatMessage)
    func onPeerListUpdated(_ peers: [String])
    /// Установлена сквозная Noise-сессия с пиром (для индикации шифрования).
    func onNoiseSession(_ peerID: String, established: Bool)
    func onError(_ error: Error)
    func onLog(_ message: String)
}

public extension MeshChatListener {
    func onPeerListUpdated(_ peers: [String]) {}
    func onNoiseSession(_ peerID: String, established: Bool) {}
    func onLog(_ message: String) {}
}

/// Реальный mesh-чат поверх портированного из bitchat-iOS BLE-стека (`BLEService`).
/// GATT-соединения, фрагментация, store-and-forward, релей с TTL, дедупликация —
/// то же ядро, что и на Android, протокол-совместимое.
public final class MeshChatSdk: NSObject {

    public weak var listener: MeshChatListener?

    private let ble: BLEService
    /// Тот же менеджер идентичности, что использует BLEService — для verified-отпечатков.
    private let identityManager: SecureIdentityStateManager
    private var started = false
    private var nickname: String
    /// Показывать локальное уведомление о входящем, когда приложение в фоне.
    /// Меняется в рантайме через setNotificationsEnabled.
    private var notifyInBackground: Bool

    /// Включить/выключить локальные уведомления о входящих сообщениях.
    public func setNotificationsEnabled(_ enabled: Bool) {
        notifyInBackground = enabled
        if enabled {
            UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { _, _ in }
        }
    }

    public var myPeerID: String { ble.myPeerID.id }

    public init(listener: MeshChatListener?, nickname: String? = nil, notifyInBackground: Bool = true) {
        self.listener = listener
        self.nickname = nickname ?? "ios-user"
        self.notifyInBackground = notifyInBackground
        let keychain = KeychainManager()
        let idBridge = NostrIdentityBridge(keychain: keychain)
        let identityManager = SecureIdentityStateManager(keychain)
        self.identityManager = identityManager
        self.ble = BLEService(keychain: keychain, idBridge: idBridge, identityManager: identityManager)
        super.init()
        self.ble.delegate = self
        // Сообщаем наверх об установлении Noise-сессии (для иконки 🔒).
        self.ble.addPeerAuthenticatedHandler { [weak self] peerID, _ in
            self?.listener?.onNoiseSession(peerID.id, established: true)
        }
        if notifyInBackground {
            UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { _, _ in }
        }
    }

    /// Локальное уведомление о входящем — только когда приложение не на переднем плане.
    private func postLocalNotificationIfBackground(sender: String, content: String) {
        guard notifyInBackground else { return }
        DispatchQueue.main.async {
            guard UIApplication.shared.applicationState != .active else { return }
            let c = UNMutableNotificationContent()
            c.title = sender.isEmpty ? "Новое сообщение" : sender
            c.body = content
            c.sound = .default
            let req = UNNotificationRequest(identifier: UUID().uuidString, content: c, trigger: nil)
            UNUserNotificationCenter.current().add(req, withCompletionHandler: nil)
        }
    }

    public func setNickname(_ nickname: String) {
        self.nickname = nickname
        ble.setNickname(nickname)
    }

    /// Режим экономии батареи. На iOS фоновый BLE троттлится самой системой,
    /// поэтому метод сохраняет предпочтение для единообразия API (поведение
    /// в основном управляется iOS). Реальный эффект — на Android.
    public private(set) var batterySaverEnabled = false
    public func setBatterySaver(_ enabled: Bool) {
        batterySaverEnabled = enabled
    }

    public func start() {
        guard !started else { return }
        started = true
        ble.setNickname(nickname)
        ble.startServices()
        ble.sendBroadcastAnnounce()
        listener?.onLog("MeshChatSdk started (peerID=\(ble.myPeerID.id))")
    }

    public func stop() {
        guard started else { return }
        started = false
        ble.stopServices()
        listener?.onLog("MeshChatSdk stopped")
    }

    /// Широковещательное сообщение всем узлам сети.
    public func sendBroadcastMessage(_ text: String) {
        guard ensureStarted() else { return }
        ble.sendMessage(text, mentions: [])
        echoOwnMessage(text, isPrivate: false)
    }

    /// Приватное сообщение конкретному узлу по его peerID.
    public func sendPrivateMessage(_ text: String, to recipientPeerID: String, recipientNickname: String) {
        guard ensureStarted() else { return }
        let messageID = UUID().uuidString
        ble.sendPrivateMessage(text, to: PeerID(str: recipientPeerID), recipientNickname: recipientNickname, messageID: messageID)
        echoOwnMessage(text, isPrivate: true)
    }

    public func getPeerNicknames() -> [String: String] {
        var result: [String: String] = [:]
        for (peer, nick) in ble.getPeerNicknames() { result[peer.id] = nick }
        return result
    }

    // MARK: - Сквозное шифрование (Noise) и верификация пиров

    /// Инициировать Noise-handshake с пиром (нужно перед первым приватным сообщением:
    /// sendPrivateMessage работает fire-and-forget и без сессии сообщение не уйдёт).
    public func startHandshake(_ peerID: String) {
        guard ensureStarted() else { return }
        ble.triggerHandshake(with: PeerID(str: peerID))
    }

    /// Есть ли установленная (зашифрованная) Noise-сессия с пиром.
    public func hasSession(_ peerID: String) -> Bool {
        if case .established = ble.getNoiseSessionState(for: PeerID(str: peerID)) {
            return true
        }
        return false
    }

    /// Отпечаток ключа пира (для сверки вне сети); nil, пока нет сессии.
    public func getPeerFingerprint(_ peerID: String) -> String? {
        return ble.getFingerprint(for: PeerID(str: peerID))
    }

    /// Наш собственный отпечаток ключа.
    public func getMyFingerprint() -> String? {
        return ble.noiseIdentityFingerprint()
    }

    /// Помечен ли пир доверенным (verified) — по отпечатку, хранится локально.
    public func isPeerVerified(_ peerID: String) -> Bool {
        guard let fp = ble.getFingerprint(for: PeerID(str: peerID)) else { return false }
        return identityManager.isVerified(fingerprint: fp)
    }

    /// Пометить/снять пир доверенным (после сверки отпечатков вне сети).
    public func setPeerVerified(_ peerID: String, _ verified: Bool) {
        guard let fp = ble.getFingerprint(for: PeerID(str: peerID)) else { return }
        identityManager.setVerified(fingerprint: fp, verified: verified)
    }

    /// Никнейм пира по peerID (или nil).
    public func getPeerNickname(_ peerID: String) -> String? {
        return getPeerNicknames()[peerID]
    }

    public func debugStatus() -> String {
        "peerID=\(ble.myPeerID.id) started=\(started) peers=\(ble.getPeerNicknames().keys.map { $0.id })"
    }

    // MARK: - Private

    private func ensureStarted() -> Bool {
        if !started {
            listener?.onError(NSError(domain: "MeshChatSdk", code: 1,
                                      userInfo: [NSLocalizedDescriptionKey: "MeshChatSdk is not started"]))
            return false
        }
        return true
    }

    /// Локальное эхо: транспорт не возвращает собственное сообщение через делегат,
    /// поэтому показываем его отправителю сами.
    private func echoOwnMessage(_ text: String, isPrivate: Bool) {
        let msg = MeshChatMessage(
            id: UUID().uuidString,
            senderPeerID: ble.myPeerID.id,
            sender: nickname,
            content: text,
            timestamp: Date().timeIntervalSince1970 * 1000,
            isPrivate: isPrivate,
            isRelay: false
        )
        listener?.onMessageReceived(msg)
    }
}

// MARK: - BitchatDelegate

extension MeshChatSdk: BitchatDelegate {
    func didReceiveMessage(_ message: BitchatMessage) {
        let msg = MeshChatMessage(
            id: message.id,
            senderPeerID: message.senderPeerID?.id,
            sender: message.sender,
            content: message.content,
            timestamp: message.timestamp.timeIntervalSince1970 * 1000,
            isPrivate: message.isPrivate,
            isRelay: message.isRelay
        )
        listener?.onMessageReceived(msg)
        postLocalNotificationIfBackground(sender: message.sender, content: message.content)
    }

    func didReceivePublicMessage(from peerID: PeerID, nickname: String, content: String, timestamp: Date, messageID: String?) {
        let msg = MeshChatMessage(
            id: messageID ?? UUID().uuidString,
            senderPeerID: peerID.id,
            sender: nickname,
            content: content,
            timestamp: timestamp.timeIntervalSince1970 * 1000,
            isPrivate: false,
            isRelay: false
        )
        listener?.onMessageReceived(msg)
        postLocalNotificationIfBackground(sender: nickname, content: content)
    }

    func didUpdatePeerList(_ peers: [PeerID]) {
        listener?.onPeerListUpdated(peers.map { $0.id })
    }

    func didConnectToPeer(_ peerID: PeerID) {}
    func didDisconnectFromPeer(_ peerID: PeerID) {}
    func isFavorite(fingerprint: String) -> Bool { false }
    func didUpdateMessageDeliveryStatus(_ messageID: String, status: DeliveryStatus) {}
    func didReceiveNoisePayload(from peerID: PeerID, type: NoisePayloadType, payload: Data, timestamp: Date) {}
    func didUpdateBluetoothState(_ state: CBManagerState) {}
}

File: android/src/main/java/com/example/meshchat/sdk/MeshChatSdk.kt
package com.example.meshchat.sdk

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ProcessLifecycleOwner
import com.bitchat.android.identity.SecureIdentityStateManager
import com.bitchat.android.mesh.BluetoothMeshDelegate
import com.bitchat.android.mesh.BluetoothMeshService
import com.bitchat.android.model.BitchatMessage
import com.bitchat.android.service.MeshServiceHolder
import com.bitchat.android.services.NicknameProvider
import com.bitchat.android.ui.debug.DebugPreferenceManager
import java.util.Date

/**
 * Сообщение mesh-чата, отдаваемое наверх (в React Native).
 *
 * Адресация идёт по строковому peerID, что позволяет поддержать приватные
 * сообщения, подтверждения доставки и прочтения.
 */
data class MeshChatMessage(
    val id: String,
    val senderPeerID: String?,
    val sender: String,
    val content: String,
    val timestamp: Long,
    val isPrivate: Boolean,
    val channel: String?,
    val isRelay: Boolean,
)

interface MeshChatListener {
    fun onMessageReceived(message: MeshChatMessage)
    fun onPeerListUpdated(peers: List<String>) {}
    fun onDeliveryAck(messageID: String, peerID: String) {}
    fun onReadReceipt(messageID: String, peerID: String) {}
    /** Установлена/сброшена сквозная Noise-сессия с пиром (для индикации шифрования). */
    fun onNoiseSession(peerID: String, established: Boolean) {}
    fun onError(error: Throwable)
    fun onLog(message: String) {}
}

/**
 * Реальный mesh-чат поверх портированного из bitchat BLE-mesh стека.
 *
 * Транспорт устанавливает настоящие GATT-соединения (server + client), режет
 * крупные пакеты на фрагменты (MTU 512), хранит сообщения для офлайн-пиров
 * (store-and-forward) и ретранслирует пакеты по сети с TTL и дедупликацией.
 */
class MeshChatSdk(
    context: Context,
    private val listener: MeshChatListener,
    nickname: String? = null,
    /**
     * Держать ли mesh живым в фоне через foreground-сервис с постоянным
     * уведомлением. По умолчанию включено — иначе Android усыпит процесс и
     * сеть перестанет работать при свёрнутом приложении.
     */
    private val keepAliveInBackground: Boolean = true,
    /**
     * Показывать ли локальное уведомление о входящем сообщении, когда приложение
     * в фоне (в foreground UI и так показывает сообщение в списке).
     * Можно менять в рантайме через [setNotificationsEnabled].
     */
    notifyInBackground: Boolean = true,
) {

    companion object {
        private const val TAG = "MeshChatSdk"
        private const val GOSSIP_OWNER = "MeshChatSdk"
    }

    /** Текущее состояние пуш-уведомлений о входящих (изменяемое). */
    @Volatile
    private var notifyInBackground: Boolean = notifyInBackground

    /** Включить/выключить локальные уведомления о входящих сообщениях. */
    fun setNotificationsEnabled(enabled: Boolean) {
        notifyInBackground = enabled
        Log.i(TAG, "notifyInBackground=$enabled")
    }

    private val appContext = context.applicationContext

    @Volatile
    private var started = false

    /** В фоне или нет — для решения, слать ли локальное уведомление. */
    @Volatile
    private var appInForeground = true

    init {
        // Включаем персистентность настроек/идентичности (best-effort).
        try { DebugPreferenceManager.init(appContext) } catch (_: Exception) {}
        nickname?.let { NicknameProvider.setNickname(it) }

        // Следим за foreground/background всего процесса (на главном потоке).
        Handler(Looper.getMainLooper()).post {
            try {
                appInForeground = ProcessLifecycleOwner.get().lifecycle.currentState
                    .isAtLeast(Lifecycle.State.STARTED)
                ProcessLifecycleOwner.get().lifecycle.addObserver(
                    LifecycleEventObserver { _: LifecycleOwner, event: Lifecycle.Event ->
                        when (event) {
                            Lifecycle.Event.ON_START -> appInForeground = true
                            Lifecycle.Event.ON_STOP -> appInForeground = false
                            else -> {}
                        }
                    }
                )
            } catch (t: Throwable) {
                Log.w(TAG, "Failed to observe process lifecycle: ${t.message}")
            }
        }
    }

    private val meshService = BluetoothMeshService(appContext)

    // Постоянное хранилище доверенных (verified) отпечатков — тот же EncryptedSharedPreferences,
    // что использует mesh-ядро (читает/пишет ключ verified_fingerprints).
    private val identityStore = SecureIdentityStateManager(appContext)

    init {
        // Сообщаем наверх, когда установлена Noise-сессия с пиром (для иконки 🔒).
        meshService.setOnNoiseSessionEstablished { peerID ->
            Log.i(TAG, "🔒 Noise session established with $peerID")
            listener.onNoiseSession(peerID, true)
        }
        meshService.delegate = object : BluetoothMeshDelegate {
            override fun didReceiveMessage(message: BitchatMessage) {
                Log.i(TAG, "📥 RX message from=${message.senderPeerID} private=${message.isPrivate} " +
                    "relay=${message.isRelay} content='${message.content.take(60)}'")
                listener.onMessageReceived(message.toMeshChatMessage())
                // В фоне JS спит — показываем нативное уведомление о входящем.
                if (notifyInBackground && !appInForeground) {
                    try {
                        MeshNotifier.notifyMessage(
                            appContext,
                            sender = message.sender,
                            content = message.content,
                            threadKey = message.senderPeerID ?: message.sender
                        )
                    } catch (t: Throwable) {
                        Log.w(TAG, "notifyMessage failed: ${t.message}")
                    }
                }
            }

            override fun didUpdatePeerList(peers: List<String>) {
                Log.i(TAG, "👥 peer list updated: count=${peers.size} peers=$peers")
                listener.onPeerListUpdated(peers)
            }

            override fun didReceiveChannelLeave(channel: String, fromPeer: String) {}

            override fun didReceiveDeliveryAck(messageID: String, recipientPeerID: String) {
                listener.onDeliveryAck(messageID, recipientPeerID)
            }

            override fun didReceiveReadReceipt(messageID: String, recipientPeerID: String) {
                listener.onReadReceipt(messageID, recipientPeerID)
            }

            override fun didReceiveVerifyChallenge(peerID: String, payload: ByteArray, timestampMs: Long) {}
            override fun didReceiveVerifyResponse(peerID: String, payload: ByteArray, timestampMs: Long) {}

            // Каналы с паролем — это фаза 2 (вместе с шифрованием), пока не расшифровываем.
            override fun decryptChannelMessage(encryptedContent: ByteArray, channel: String): String? = null

            override fun getNickname(): String? =
                NicknameProvider.getNickname(appContext, meshService.myPeerID)

            override fun isFavorite(peerID: String): Boolean = false
        }
    }

    /** Идентификатор этого узла в mesh-сети (стабильный, привязан к ключу). */
    val myPeerID: String get() = meshService.myPeerID

    fun setNickname(nickname: String?) {
        NicknameProvider.setNickname(nickname)
    }

    /**
     * Режим экономии батареи. При включении mesh в фоне работает реже
     * (POWER_SAVER): меньше расход, но доставка в фоне менее оперативна.
     * По умолчанию выключен (в фоне держим BALANCED).
     */
    fun setBatterySaver(enabled: Boolean) {
        com.bitchat.android.mesh.PowerManager.setBatterySaver(enabled)
        Log.i(TAG, "batterySaver=$enabled")
    }

    fun isBatterySaverEnabled(): Boolean =
        com.bitchat.android.mesh.PowerManager.batterySaverEnabled

    fun start() {
        if (started) {
            Log.w(TAG, "start() ignored: already started")
            return
        }
        started = true
        try {
            Log.i(TAG, "▶️ start() peerID=${meshService.myPeerID} nickname=${NicknameProvider.getNickname(appContext, meshService.myPeerID)}")
            MeshServiceHolder.meshService = meshService
            meshService.startServices()
            MeshServiceHolder.startSharedGossip(GOSSIP_OWNER)
            // Сразу анонсируем себя, чтобы соседние узлы нас увидели.
            meshService.sendBroadcastAnnounce()
            // Foreground-сервис: держим процесс живым в фоне.
            if (keepAliveInBackground) {
                try { MeshForegroundService.start(appContext) } catch (t: Throwable) {
                    Log.w(TAG, "Failed to start foreground service: ${t.message}")
                }
            }
            Log.i(TAG, "✅ start() done. ${transportStatus()}")
            listener.onLog("MeshChatSdk started (peerID=${meshService.myPeerID})")
        } catch (t: Throwable) {
            started = false
            Log.e(TAG, "❌ Failed to start mesh service", t)
            listener.onError(t)
        }
    }

    /** Короткая сводка состояния транспорта — для диагностики. */
    private fun transportStatus(): String = try {
        "activePeers=${meshService.getActivePeerCount()} peers=${meshService.getPeerNicknames().keys}"
    } catch (t: Throwable) {
        "status unavailable: ${t.message}"
    }

    /** Полный отладочный статус транспорта (можно дёргать из RN). */
    fun debugStatus(): String = try {
        meshService.getDebugStatus()
    } catch (t: Throwable) {
        "debugStatus unavailable: ${t.message}"
    }

    fun stop() {
        if (!started) return
        started = false
        try {
            if (keepAliveInBackground) {
                try { MeshForegroundService.stop(appContext) } catch (_: Throwable) {}
            }
            MeshServiceHolder.stopSharedGossip(GOSSIP_OWNER)
            meshService.stopServices()
            if (MeshServiceHolder.meshService === meshService) {
                MeshServiceHolder.meshService = null
            }
            listener.onLog("MeshChatSdk stopped")
        } catch (t: Throwable) {
            Log.e(TAG, "Failed to stop mesh service", t)
            listener.onError(t)
        }
    }

    /** Широковещательное сообщение всем узлам сети. */
    fun sendBroadcastMessage(text: String) {
        Log.i(TAG, "📤 sendBroadcast len=${text.length} started=$started ${transportStatus()}")
        if (!ensureStarted()) return
        try {
            meshService.sendMessage(text)
            Log.i(TAG, "📤 sendBroadcast -> meshService.sendMessage() returned ok")
            // Локальное эхо: транспорт только передаёт пакет в сеть и не возвращает
            // собственное сообщение через didReceiveMessage, поэтому показываем его
            // отправителю сами — иначе кажется, что «сообщение не отправилось».
            echoOwnMessage(text, isPrivate = false, recipientNickname = null)
        } catch (t: Throwable) {
            Log.e(TAG, "❌ Failed to send broadcast message", t)
            listener.onError(t)
        }
    }

    /** Приватное сообщение конкретному узлу по его peerID. */
    fun sendPrivateMessage(text: String, recipientPeerID: String, recipientNickname: String) {
        Log.i(TAG, "📤 sendPrivate to=$recipientPeerID len=${text.length} started=$started ${transportStatus()}")
        if (!ensureStarted()) return
        try {
            meshService.sendPrivateMessage(text, recipientPeerID, recipientNickname)
            Log.i(TAG, "📤 sendPrivate -> meshService.sendPrivateMessage() returned ok")
            echoOwnMessage(text, isPrivate = true, recipientNickname = recipientNickname)
        } catch (t: Throwable) {
            Log.e(TAG, "❌ Failed to send private message", t)
            listener.onError(t)
        }
    }

    // --- Сквозное шифрование (Noise) и верификация пиров ---

    /**
     * Инициировать Noise-handshake с пиром. Нужно вызвать перед первой отправкой
     * приватного сообщения: `sendPrivateMessage` работает по принципу fire-and-forget
     * и, если сессии ещё нет, сообщение НЕ отправится (только запустится handshake).
     * Об установлении сессии придёт onNoiseSession(peerID, true).
     */
    fun startHandshake(peerID: String) {
        if (!ensureStarted()) return
        Log.i(TAG, "🤝 startHandshake with $peerID")
        meshService.initiateNoiseHandshake(peerID)
    }

    /** Есть ли установленная (зашифрованная) Noise-сессия с пиром. */
    fun hasSession(peerID: String): Boolean = try {
        meshService.hasEstablishedSession(peerID)
    } catch (_: Exception) { false }

    /** Отпечаток ключа пира (SHA-256 статического Noise-ключа) — для сверки вне сети. */
    fun getPeerFingerprint(peerID: String): String? = try {
        meshService.getPeerFingerprint(peerID)
    } catch (_: Exception) { null }

    /** Наш собственный отпечаток ключа (его сверяет собеседник). */
    fun getMyFingerprint(): String? = try {
        meshService.getIdentityFingerprint()
    } catch (_: Exception) { null }

    /** Помечен ли пир как доверенный (verified) — по отпечатку, хранится локально. */
    fun isPeerVerified(peerID: String): Boolean {
        return try {
            val fp = meshService.getPeerFingerprint(peerID) ?: return false
            identityStore.isVerifiedFingerprint(fp)
        } catch (_: Exception) {
            false
        }
    }

    /** Пометить/снять пир как доверенный (после сверки отпечатков вне сети). */
    fun setPeerVerified(peerID: String, verified: Boolean) {
        try {
            val fp = meshService.getPeerFingerprint(peerID)
            if (fp == null) {
                Log.w(TAG, "setPeerVerified: нет отпечатка для $peerID (сессия не установлена?)")
                return
            }
            identityStore.setVerifiedFingerprint(fp, verified)
            Log.i(TAG, "🛡️ peer $peerID verified=$verified (fp=${fp.take(16)}…)")
        } catch (t: Throwable) {
            Log.e(TAG, "setPeerVerified failed", t)
        }
    }

    /** Никнейм пира по его peerID (или null, если неизвестен). */
    fun getPeerNickname(peerID: String): String? = getPeerNicknames()[peerID]

    private fun echoOwnMessage(text: String, isPrivate: Boolean, recipientNickname: String?) {
        Log.i(TAG, "🪞 echo own message to listener (private=$isPrivate)")
        val myNick = NicknameProvider.getNickname(appContext, meshService.myPeerID)
        val echo = BitchatMessage(
            sender = myNick,
            content = text,
            timestamp = Date(),
            senderPeerID = meshService.myPeerID,
            isPrivate = isPrivate,
            recipientNickname = recipientNickname,
        )
        listener.onMessageReceived(echo.toMeshChatMessage())
    }

    fun getPeerNicknames(): Map<String, String> = try {
        meshService.getPeerNicknames()
    } catch (_: Exception) {
        emptyMap()
    }

    private fun ensureStarted(): Boolean {
        if (!started) {
            Log.e(TAG, "⛔ send blocked: MeshChatSdk is not started")
            listener.onError(IllegalStateException("MeshChatSdk is not started"))
            return false
        }
        return true
    }

    private fun BitchatMessage.toMeshChatMessage() = MeshChatMessage(
        id = id,
        senderPeerID = senderPeerID,
        sender = sender,
        content = content,
        timestamp = timestamp.time,
        isPrivate = isPrivate,
        channel = channel,
        isRelay = isRelay,
    )
}
